Documentation/botr/shared-generics.md

```diff
@@ -0,0 +1,185 @@
+
+Given an input signature, slot assignment is performed with the following algorithm:
+
+```
+Begin with slot = 0
+Foreach entry in dictionary layout
+    If entry.signature != NULL
+        If entry.signature == inputSignature
+            return slot
+        EndIf
+    Else
+        entry.signature = inputSignature
+        return slot
+    EndIf
+    slot++
+EndForeach
+```
+
+So what happens when the above algorithm runs, but no existing slot with the same signature is found, and we're out of 'unassigned' slots? This is where the dynamic dictionary expansion kicks in to resize the layout by adding more slots to it, and resizing all dictionaries associated with this layout.
+
+# Dynamic Dictionary Expansion
+
+### History
+
+Before the dynamic dictionary expansion feature, dictionary layouts were organized into buckets (a linked list of fixed-size `DictionaryLayout` structures). The size of the initial layout bucket was always fixed to some number which was computed based on some heuristics for generic types, and always fixed to 4 slots for generic methods. The generic types and methods also had fixed-size generic dictionaries which could be used for lookups (also known as "fast lookup slots").
+
+When a bucket gets filled with entries, we would just allocate a new `DictionaryLayout` bucket, and add it to the list. The problem however is that we couldn't resize the generic dictionaries of types or methods, because they have already been allocated with a fixed size, and the JIT does not support generating instructions that could indirect into a linked-list of dictionaries. Given that limitation, we could only lookup a generic dictionary for a fixed number of values (the ones associated with the entries of the first `DictionaryLayout` bucket), and were forced to go through a slower runtime helper for additional lookups.
+
+This was acceptable, until we introduced the [ReadyToRun](https://github.com/dotnet/coreclr/blob/master/Documentation/botr/readytorun-overview.md) and the Tiered Compilation technologies. Slots were getting assigned quickly when used by ReadyToRun code, and when the runtime decided re-jitted certain methods for better performance, it could not in some cases find any remaining "fast lookup slots", and was forced to generate code that goes through the slower runtime helpers. This ended up hurting performance in some scenarios, and a decision was made to not use the fast lookup slots for ReadyToRun code, and instead keep them reserved for re-jitted code. This decision however hurt the ReadyToRun performance, but it was a necessary compromise since we cared more about re-jitted code throughput over R2R throughput.
```

Comment:
I would not says that it was acceptable. We had number of customers running into the performance cliff, for example https://github.com/dotnet/coreclr/issues/55 . I had to answer issues like this many times. We actually did tweaked the heuristic several times to make it work better.

The fundamental reason why we have not done anything about this was fragile NGen. It was not realistic to build the tracing structure like you have built and make it work well with fragile NGen.

Reply:
Thanks for the clarification Jan